feat(pypi/parse_requirements): get dists by version when no hash provied

Yanpei-Wang opened the pull request:
This pull request modifies the SimpleAPI HTML parsing to add a new
field where we can get the `sha256` values by package version. This
allows us to very easily fallback to all packages of a particular
version when using `experimental_index_url` if the hashes are not
specified.

The code deciding which packages to query the SimpleAPI for has been
also modified to only omit queries for packages that are included via
direct URL references.

If we fail to get the data from the SimpleAPI, we will fallback to
`pip` and try to install it via the legacy behaviour.

Fixes #2023
Work towards #260
Work towards #1357
Work towards #2363

aignas commented on python/private/pypi/parse_requirements.bzl:
```diff
@@ -315,26 +315,46 @@ def _add_dists(*, requirement, index_urls, logger = None):
+
+            if logger:
+                logger.warn(lambda: "Could not find a whl or an sdist with sha256={}".format(sha256))
+    else:
+        # If no SHA256s provided, try to find distributions by version
+        version = requirement.srcs.version
+        if version:
```
```suggestion
    elif requirement.srcs.version:
        # If no SHA256s provided, try to find distributions by version
```

could be a simple way to define it and it avoids having the indentation.

aignas commented on python/private/pypi/parse_requirements.bzl:
```diff
@@ -315,26 +315,46 @@ def _add_dists(*, requirement, index_urls, logger = None):
+        version = requirement.srcs.version
+        if version:
+            # Look for wheels matching the version
+            for whl in index_urls.whls.values():
```
Ideally we could have `index_urls.shas_by_version` which we could use instead and unify the codepaths.

aignas commented on tests/pypi/parse_requirements/parse_requirements_tests.bzl:
```diff
@@ -563,6 +567,62 @@ def _test_different_package_version(env):
+                whls = [struct(
+                    url = "https://example.org/foo-0.0.4.whl",
+                    filename = "foo-0.0.4.whl",
+                    sha256 = "",
```
We should ensure that `whl_library` reports a warning if the sha256 is not specified, but the URL is. We should ask the user to add the sha256 to their manifest. We can print the sha256 value.

aignas commented on tests/pypi/parse_requirements/parse_requirements_tests.bzl:
```diff
@@ -563,6 +567,62 @@ def _test_different_package_version(env):
+        ctx = _mock_ctx(),
+        requirements_by_platform = {
+            "requirements_optional_hash": ["linux_x86_64"],
+        },
```
Could you add a test where the `get_index_urls` is set so that we are testing the newly added code?

aignas reviewed (changes requested):
Thanks for the PR. For this to be more robust and performant, it would be great to add an extra field that is the result of SimpleAPI parsing where we store sha256s by version of the package in the index.

The code is here:
https://github.com/bazel-contrib/rules_python/blob/main/python/private/pypi/parse_simpleapi_html.bzl#L96

aignas commented:
OK, I have updated the implementation based on my own feedback and updated the
docs. Could you please test if it works in your environment?

Yanpei-Wang commented:
> OK, I have updated the implementation based on my own feedback and updated the docs. Could you please test if it works in your environment?

Yanpei-Wang commented:
> OK, I have updated the implementation based on my own feedback and updated the docs. Could you please test if it works in your environment?

Yanpei-Wang commented:
> OK, I have updated the implementation based on my own feedback and updated the docs. Could you please test if it works in your environment?

Thank you for your help. It's my first time doing this. I mistakenly closed it but reopened it.

Yanpei-Wang commented:
> OK, I have updated the implementation based on my own feedback and updated the docs. Could you please test if it works in your environment?

Thanks for the updates and feedback! I’m still new to this, so I really appreciate your help. I’ve checked the updated implementation and docs, and I’ve learned a lot. I tested it in my environment, and it works great.
